File: tel/receiver.py
"""Dependency-free OTLP receiver, used when Docker is unavailable.

Accepts OTLP over **HTTP/JSON only** (`OTEL_EXPORTER_OTLP_PROTOCOL=http/json`)
and appends each request verbatim to the same newline-delimited JSON files the
collector writes, so the analysis layer cannot tell the difference.

This is a fallback. The OpenTelemetry Collector is the supported path: it also
speaks gRPC and http/protobuf, handles batching and file rotation, and is far
better tested.
"""
from __future__ import annotations

import datetime as _dt
import gzip
import json
import signal
import sys
import threading
from http.server import BaseHTTPRequestHandler, ThreadingHTTPServer
from pathlib import Path

from . import config

_LOCK = threading.Lock()

PATHS = {
    "/v1/logs": ("logs", "resourceLogs"),
    "/v1/metrics": ("metrics", "resourceMetrics"),
    "/v1/traces": ("traces", "resourceSpans"),
}


def _append(signal_name: str, payload: dict) -> None:
    out = Path(config.RAW_DIR) / f"{signal_name}.jsonl"
    with _LOCK:
        with out.open("a", encoding="utf-8") as fh:
            fh.write(json.dumps(payload, separators=(",", ":")) + "\n")


class Handler(BaseHTTPRequestHandler):
    protocol_version = "HTTP/1.1"
    server_version = "claude-telemetry-receiver"

    def log_message(self, fmt, *args):  # keep stdout useful
        pass

    def _reply(self, code: int, body: bytes = b"{}") -> None:
        self.send_response(code)
        self.send_header("Content-Type", "application/json")
        self.send_header("Content-Length", str(len(body)))
        self.end_headers()
        self.wfile.write(body)

    def do_GET(self):
        if self.path.rstrip("/") in ("/health", ""):
            self._reply(200, b'{"status":"Server available"}')
        else:
            self._reply(404, b'{"error":"not found"}')

    def do_POST(self):
        route = PATHS.get(self.path.split("?")[0])
        if not route:
            self._reply(404, b'{"error":"unsupported signal path"}')
            return
        signal_name, root_key = route
        try:
            length = int(self.headers.get("Content-Length") or 0)
            raw = self.rfile.read(length) if length else b""
            if (self.headers.get("Content-Encoding") or "").lower() == "gzip":
                raw = gzip.decompress(raw)
            ctype = (self.headers.get("Content-Type") or "").lower()
            if "json" not in ctype:
                self._reply(415, json.dumps({
                    "error": "this fallback receiver accepts OTLP/JSON only; "
                             "set OTEL_EXPORTER_OTLP_PROTOCOL=http/json, or run "
                             "the collector with `telemetry start`"}).encode())
                return
            payload = json.loads(raw.decode("utf-8"))
        except Exception as exc:
            self._reply(400, json.dumps({"error": str(exc)}).encode())
            return

        if root_key not in payload:
            payload = {root_key: payload.get(root_key, [])}
        _append(signal_name, payload)
        self._reply(200)


def serve(port: int | None = None) -> None:
    config.ensure_dirs()
    port = port or config.OTLP_HTTP_PORT
    httpd = ThreadingHTTPServer(("127.0.0.1", port), Handler)

    def _stop(*_):
        threading.Thread(target=httpd.shutdown, daemon=True).start()

    signal.signal(signal.SIGTERM, _stop)
    signal.signal(signal.SIGINT, _stop)
    started = _dt.datetime.now(_dt.timezone.utc).isoformat(timespec="seconds")
    print(f"[{started}] fallback OTLP/JSON receiver on http://127.0.0.1:{port} "
          f"-> {config.RAW_DIR}", flush=True)
    try:
        httpd.serve_forever(poll_interval=0.5)
    finally:
        httpd.server_close()
        print("receiver stopped", flush=True)


if __name__ == "__main__":
    serve(int(sys.argv[1]) if len(sys.argv) > 1 else None)

File: src/receiver.ts
/** The OTLP receiver: the only thing standing between Claude Code and disk.
 *
 * Accepts OTLP over HTTP/JSON on loopback and appends each request verbatim to
 * newline-delimited JSON files under ~/.telemetry/raw. That is the whole job.
 * Everything downstream reads those files, so the receiver never needs to know
 * what any of it means, and a crash here can lose at most one in-flight batch.
 *
 * Node's standard library only - no collector, no container, no daemon
 * manager. `telemetry start` runs this in the background.
 */
import { appendFileSync, renameSync, statSync } from 'node:fs'
import {
  createServer,
  type IncomingMessage,
  type ServerResponse,
} from 'node:http'
import { join } from 'node:path'
import { gunzipSync } from 'node:zlib'
import * as config from './config.js'

/** Roll a raw file over at this size and start a fresh one. Nothing deletes
 * the rolled files: `analyse` reads them all, and pruning is the operator's
 * call. */
const MAX_BYTES = 64 * 1024 * 1024

const PATHS: Record<string, [string, string]> = {
  '/v1/logs': ['logs', 'resourceLogs'],
  '/v1/metrics': ['metrics', 'resourceMetrics'],
  '/v1/traces': ['traces', 'resourceSpans'],
}

function timestampSuffix(): string {
  const d = new Date()
  const p = (n: number) => String(n).padStart(2, '0')
  return (
    `${d.getFullYear()}-${p(d.getMonth() + 1)}-${p(d.getDate())}` +
    `T${p(d.getHours())}-${p(d.getMinutes())}-${p(d.getSeconds())}`
  )
}

/** Move a full file aside under a timestamped name. */
function rotate(out: string): void {
  try {
    const dot = out.lastIndexOf('.')
    const stem = out.slice(0, dot)
    const ext = out.slice(dot)
    renameSync(out, `${stem}-${timestampSuffix()}-size${ext}`)
  } catch {
    /* somebody else rotated it, or the directory is gone */
  }
}

function append(signalName: string, payload: unknown): void {
  const out = join(config.RAW_DIR, `${signalName}.jsonl`)
  // Node runs this on one thread, so the append is already serialised; the
  // only ordering that matters is rotate-then-write.
  try {
    if (statSync(out).size >= MAX_BYTES) rotate(out)
  } catch {
    /* the file does not exist yet, which is the common case on first write */
  }
  appendFileSync(out, JSON.stringify(payload) + '\n', 'utf8')
}

function reply(res: ServerResponse, code: number, body = '{}'): void {
  const buf = Buffer.from(body, 'utf8')
  res.writeHead(code, {
    'Content-Type': 'application/json',
    'Content-Length': String(buf.length),
  })
  res.end(buf)
}

function readBody(req: IncomingMessage): Promise<Buffer> {
  return new Promise((resolve, reject) => {
    const chunks: Buffer[] = []
    req.on('data', (chunk: Buffer) => chunks.push(chunk))
    req.on('end', () => resolve(Buffer.concat(chunks)))
    req.on('error', reject)
  })
}

async function handleRequest(
  req: IncomingMessage,
  res: ServerResponse,
): Promise<void> {
  const path = (req.url || '').split('?')[0]!

  if (req.method === 'GET') {
    const normalized = path.replace(/\/+$/, '')
    if (normalized === '/health' || normalized === '') {
      reply(res, 200, '{"status":"Server available"}')
    } else {
      reply(res, 404, '{"error":"not found"}')
    }
    return
  }

  if (req.method !== 'POST') {
    reply(res, 404, '{"error":"not found"}')
    return
  }

  const route = PATHS[path]
  if (!route) {
    reply(res, 404, '{"error":"unsupported signal path"}')
    return
  }
  const [signalName, rootKey] = route

  let payload: Record<string, unknown>
  try {
    let raw = await readBody(req)
    if ((req.headers['content-encoding'] || '').toLowerCase() === 'gzip') {
      raw = gunzipSync(raw)
    }
    const ctype = (req.headers['content-type'] || '').toLowerCase()
    if (!ctype.includes('json')) {
      reply(
        res,
        415,
        JSON.stringify({
          error:
            'this receiver accepts OTLP/JSON only; set ' +
            'OTEL_EXPORTER_OTLP_PROTOCOL=http/json (run ' +
            '`telemetry install` to fix your settings)',
        }),
      )
      return
    }
    payload = JSON.parse(raw.toString('utf8'))
  } catch (exc) {
    reply(res, 400, JSON.stringify({ error: (exc as Error).message }))
    return
  }

  if (!(rootKey in payload)) {
    payload = { [rootKey]: (payload as Record<string, unknown>)[rootKey] ?? [] }
  }
  append(signalName, payload)
  reply(res, 200)
}

export function serve(port?: number): void {
  config.ensureDirs()
  const listenPort = port || config.OTLP_PORT
  const server = createServer((req, res) => {
    handleRequest(req, res).catch((exc) => {
      try {
        reply(res, 500, JSON.stringify({ error: (exc as Error).message }))
      } catch {
        /* the client hung up first */
      }
    })
  })

  const stop = () => {
    server.close(() => {
      process.stdout.write('receiver stopped\n')
      process.exit(0)
    })
  }
  process.on('SIGTERM', stop)
  process.on('SIGINT', stop)

  server.on('error', (err: NodeJS.ErrnoException) => {
    if (err.code === 'EADDRINUSE') {
      process.stderr.write(
        `port ${listenPort} is already in use - another receiver is running, ` +
          `or the previous one has not finished shutting down\n`,
      )
      process.exit(1)
    }
    throw err
  })

  server.listen(listenPort, '127.0.0.1', () => {
    const started = new Date().toISOString().replace(/\.\d{3}Z$/, '+00:00')
    process.stdout.write(
      `[${started}] OTLP/JSON receiver on http://127.0.0.1:${listenPort} -> ${config.RAW_DIR}\n`,
    )
  })
}
